dll/win32/shell32/shelldesktop/CDesktopBrowser.cpp

```diff
@@ -483,6 +483,201 @@ HANDLE WINAPI SHCreateDesktop(IShellDesktopTray *Tray)
+    LPCWSTR pszLinkPath,
+    LPCWSTR pszTargetPath OPTIONAL,
+    LPCITEMIDLIST pidlTarget OPTIONAL,
+    LPCWSTR pszArg OPTIONAL,
+    LPCWSTR pszDir OPTIONAL,
+    LPCWSTR pszIconPath OPTIONAL,
+    INT iIconNr OPTIONAL,
+    LPCWSTR pszComment OPTIONAL)
+{
+    IShellLinkW *psl = NULL;
+    IPersistFile *ppf = NULL;
+    HRESULT hr = CoInitialize(NULL);
+
+    if (FAILED_UNEXPECTEDLY(hr))
+        goto Quit;
+
+    hr = CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER,
+                          IID_PPV_ARG(IShellLinkW, &psl));
+    if (FAILED_UNEXPECTEDLY(hr))
+        goto Quit;
+
+    if (pszTargetPath)
+    {
+        hr = psl->SetPath(pszTargetPath);
+        if (FAILED_UNEXPECTEDLY(hr))
+            goto Quit;
+    }
+    else if (pidlTarget)
+    {
+        hr = psl->SetIDList(pidlTarget);
+        if (FAILED_UNEXPECTEDLY(hr))
+            goto Quit;
+    }
+    else
+    {
+        ERR("invalid argument\n");
+        hr = E_INVALIDARG;
+        goto Quit;
+    }
+
+    if (pszArg)
+        hr = psl->SetArguments(pszArg);
+
+    if (pszDir)
+        hr = psl->SetWorkingDirectory(pszDir);
+
+    if (pszIconPath)
+        hr = psl->SetIconLocation(pszIconPath, iIconNr);
+
+    if (pszComment)
+        hr = psl->SetDescription(pszComment);
+
+    hr = psl->QueryInterface(IID_PPV_ARG(IPersistFile, &ppf));
+    if (FAILED_UNEXPECTEDLY(hr))
+        goto Quit;
+
+    hr = ppf->Save(pszLinkPath, TRUE);
+    if (FAILED_UNEXPECTEDLY(hr))
+        goto Quit;
+
+Quit:
+    if (psl)
+        psl->Release();
+    if (ppf)
+        ppf->Release();
+    CoUninitialize();
+    return hr;
+}
+
+static BOOL SendToFolderHasAnyItems(void)
+{
+    WCHAR szPath[MAX_PATH];
+    SHGetSpecialFolderPathW(NULL, szPath, CSIDL_SENDTO, FALSE);
+
+    PathAppendW(szPath, L"*");
+
+    WIN32_FIND_DATAW find;
+    HANDLE hFind = FindFirstFileW(szPath, &find);
+    if (hFind == INVALID_HANDLE_VALUE)
+        return FALSE;
+
+    BOOL bFound = FALSE;
+    do
+    {
+        if (wcscmp(find.cFileName, L".") == 0 ||
+            wcscmp(find.cFileName, L"..") == 0 ||
+            _wcsicmp(find.cFileName, L"desktop.ini") == 0)
+        {
+            continue;
+        }
+
+        bFound = TRUE;
+        break;
+    } while (FindNextFileW(hFind, &find));
+
+    FindClose(hFind);
+    return bFound;
+}
+
+static BOOL
+CreateSendToMyDocuments(LPCWSTR pszSendTo)
+{
+    WCHAR szTarget[MAX_PATH], szSendToFile[MAX_PATH], szShell32[MAX_PATH];
+
+    SHGetSpecialFolderPathW(NULL, szTarget, CSIDL_MYDOCUMENTS, FALSE);
+
+    StringCbCopyW(szSendToFile, sizeof(szSendToFile), pszSendTo);
+    PathAppendW(szSendToFile, PathFindFileNameW(szTarget));
+    StringCbCatW(szSendToFile, sizeof(szSendToFile), L".lnk");
+
+    GetSystemDirectoryW(szShell32, _countof(szShell32));
+    PathAppendW(szShell32, L"shell32.dll");
+
+    HRESULT hr = CreateShellLink(szSendToFile, szTarget, NULL, NULL, NULL,
+                                 szShell32, -IDI_SHELL_MY_DOCUMENTS, NULL);
```

Comment:
Question (maybe stupid) : why "-IDI_SHELL_MY_DOCUMENTS" and not "IDI_SHELL_MY_DOCUMENTS" ?

Reply:
Negative value means icon resource ID rather than icon index.